sql/core/src/test/scala/org/apache/spark/sql/connector/MetadataColumnSuite.scala

```diff
@@ -232,4 +237,175 @@ class MetadataColumnSuite extends DatasourceV2SQLBase {
       )
     }
   }
+
+  test("SPARK-41498: Metadata column is propagated through union") {
```

Comment:
could we also add a positive example with more complicated columns?
e.g. selecting _partition and _metadata; or projecting _metadata to only select file_name so that nested column pruning kicks in?